Add form wizard utils

rufflian opened the pull request:
@long-banana here's a quick look at the form and wizard mechanics, draft ~one~two.

Uses Formik for "inline" inputs.

The `Wizard` is designed to support dynamic steps, so as a user completes prior steps we can adjust those coming up. This is done using conditionally rendered React children (for an example see https://github.com/raindrop-studios/creator-ui/blob/fe66d3c22b399b0bbef54167ceb5fdf2142c74bf/src/components/Wizard/Wizard.stories.tsx). 
Demo: https://www.loom.com/share/83e6228609094ba5a71f3e23199afcd4

Validation is done using yup

- Adds basic form and wizard utils, inputs

long-banana commented on src/assets/paste.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" viewbox="0 0 48 48"><path d="M9 42q-1.3 0-2.15-.85Q6 40.3 6 39V9q0-1.3.85-2.15Q7.7 6 9 6h10.1q.35-1.75 1.725-2.875T24 2q1.8 0 3.175 1.125Q28.55 4.25 28.9 6H39q1.3 0 2.15.85Q42 7.7 42 9v30q0 1.3-.85 2.15Q40.3 42 39 42Zm0-3h30V9h-3v4.5H12V9H9v30ZM24 9q.85 0 1.425-.575Q26 7.85 26 7q0-.85-.575-1.425Q24.85 5 24 5q-.85 0-1.425.575Q22 6.15 22 7q0 .85.575 1.425Q23. ...
```
Unless it is a rendering issue with GitHub, the canvas for this SVG is much larger than the actual content. Was that intentional?

rufflian replied:
Nope! Let me see if I can fix

long-banana commented on src/components/Wizard/FormStep.tsx:
```diff
@@ -0,0 +1,63 @@
+  status?: "future" | "inprogress" | "complete" | "error" | "skipped";
+  onPrevious?: (arg?: any) => void;
+  onNext?: (arg?: any) => void;
+  Component: React.FC<ComponentProps & SubStepProps>;
```
Where do the types `React`, and `ComponentProps` get brought into scope?

rufflian replied:
ComponentProps is passed into the step, React updated

long-banana commented on src/components/Wizard/FormStep.tsx:
```diff
@@ -0,0 +1,63 @@
+  title: string;
+  hash: string;
+  description?: React.ReactNode;
+  status?: "future" | "inprogress" | "complete" | "error" | "skipped";
```
Enum?

rufflian replied:
No longer used

long-banana commented on src/components/Wizard/FormStep.tsx:
```diff
@@ -0,0 +1,63 @@
+      justifyContent="flex-start"
+      flexDirection="column"
+    >
+      {!!previous ? (
```
Instead of doing a ternary in the return expression, what do you think about having a variable above that holds null by default, and if previous is truthy, you set it to the Button? Then this simplifies to `{ previous }` in the return statement. Slightly easier to parse than a ternary.

long-banana replied:
Or, if you really want to keep it all in the return, you could just &&

```
{ !!wallet_connected && (<Header showWalletConnect={true} /> )}
```

rufflian replied:
This has been updated to return an empty div for layout purposes in case of no previous

long-banana replied:
My primary comment was in reference to having if/else branching within the return statement. For me at least it's kinda messy and challenging to parse rather then using branching above the return with an if statement thereby keeping the return statement more explicit with less branching. You're fine to keep it how you want in this case. I know it's not uncommon with React, I just wanted to clarify my reasoning.

long-banana commented on src/components/Wizard/Inputs/FormControlBlock.tsx:
```diff
@@ -0,0 +1,27 @@
+export type FormControlBlockProps = {
+  title?: FormControlProps["label"];
+  help?: FormControlProps["caption"];
+  children: React.ReactNode;
```
Import `React` depending on the answer to my bringing into scope question in the `FormStep.tsx` file.

rufflian replied:
Updated

rufflian commented:
@long-banana I've updated the UI to use a tab layout for now, see here for a quick peek: https://www.loom.com/share/83e6228609094ba5a71f3e23199afcd4

long-banana commented on src/components/Wizard/Inputs/Integer.tsx:
```diff
@@ -0,0 +1,47 @@
+}: IntegerInputProps<FormValues>) {
+  const { value, setValue, error } = useFormField<FormValues>(name);
+  const [currentValue, setCurrentValue] = React.useState<number>(
+    value as unknown as number || min || 0
```
what is this `as unknown` voodoo you're doing here? I'm not familiar.

rufflian replied:
This was to get around the potential mismatch in types (value could be anything, but IntegerInput would only take numbers).

I've updated that here https://github.com/raindrop-studios/creator-ui/pull/1/commits/f22e89b796585fa7f1ca9bae9ad6cacbeddce535 to pass the expected type into `useFormField` and handle it in there

long-banana commented on src/components/Wizard/Inputs/Integer.tsx:
```diff
@@ -0,0 +1,47 @@
+import * as React from "react";
```
`import React, { useState } from "react";`

You did this in another file. Do you want to here also?

long-banana reviewed (approved):
Nice design!

rufflian commented:
@long-banana I've refactored the inputs etc. to use Formik (I was running into issues with the custom forms hook with repeat validation) - this is inline only, you can still use standalone steps, tidied up the janky inputs and added the "go to any step in the past" feature. Ready for re-review!

long-banana commented on src/components/Wizard/useFormWizard.ts:
```diff
@@ -0,0 +1,45 @@
+      complete();
+    }
+  }, [currentStep]);
+  const goToPrevious = () => {
```
Do you want to do any bounds checking on this function and `goToNext`?

rufflian replied:
Updated!

long-banana reviewed (approved):
Looks good!
